File: strings/01_neigh/test_neigh.py
'''
PROBLEM STATEMENT:

How many horses are there a string with multiple 'neighs' in it;
Given a string (of horse sound 'neigh'),
determine the minimum number of horses possible.
'Nneigheigh' should return 2, 'neighneigh' returns 1.
'''

from unittest import TestCase, TestResult, TestLoader
from neigh import Solution
from time import time

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


@time_tests
class TestNeigh(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_Single(self):
        sound = "neigh"
        expected = 1
        result = self.solution.min_horses(sound)

        self.assertEqual(expected, result)

    def test2_anotherSingle(self):
        sound = "neighneigh"
        expected = 1
        result = self.solution.min_horses(sound)

        self.assertEqual(expected, result)

    def test3_Double(self):
        sound = "nneigheigh"
        expected = 2
        result = self.solution.min_horses(sound)

        self.assertEqual(expected, result)

    def test4_Triple(self):
        sound = "nneingheighigh"
        expected = 3
        result = self.solution.min_horses(sound)

        self.assertEqual(expected, result)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestNeigh)
    result = CustomTestResult()
    suite.run(result)
    result.printTestTimes()

File: strings/01_neigh/test_neigh.cpp
/*
PROBLEM STATEMENT:

How many horses are there a string with multiple 'neighs' in it;
Given a string (of horse sound 'neigh'),
determine the minimum number of horses possible.
'Nneigheigh' should return 2, 'neighneigh' returns 1.
*/

#include "neigh.cpp"
#include <gtest/gtest.h>
#include <iostream>
#include <chrono>
#include <iostream>

using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestNeigh : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestNeigh, test1_Single) { 
    string sound = "neigh";
    int expected = 1;
    int result = solution->min_horses(sound);

    ASSERT_EQ(expected, result);
}

TEST_F(TestNeigh, test2_anotherSingle) { 
    string sound = "neighneigh";
    int expected = 1;
    int result = solution->min_horses(sound);

    ASSERT_EQ(expected, result);
}

TEST_F(TestNeigh, test3_Double) { 
    string sound = "nneigheigh";
    int expected = 2;
    int result = solution->min_horses(sound);

    ASSERT_EQ(expected, result);
}

TEST_F(TestNeigh, test4_Triple) { 
    string sound = "nneingheighigh";
    int expected = 3;
    int result = solution->min_horses(sound);

    ASSERT_EQ(expected, result);
}


int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}